### PR TITLE
More re-structuring of the file system watching code

### PR DESCRIPTION
**Pull request by wolfs**
The main change is switching to one global VFS root which holds the whole state, while keeping around to `VirtualFileSystemService`s for changing the global root.

The global root now is of type `VfsRootReference`, which allows obtaining a read-only view of the VFS via `get()` and allows changing the VFS via an `update()` method. The clients of this API do not need to know anything about diff listeners any more, since those are hidden behind `VfsRoot`.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java [hunk 1]
```diff
@@ -190,20 +189,19 @@ WatchFilter createWatchFilter(GlobalCacheLocations globalCacheLocations) {
...
+        VfsRootReference createVfsRootReference(FileSystem fileSystem, SnapshotHierarchy.UpdateFunctionRunner updateFunctionRunner) {
             CaseSensitivity caseSensitivity = fileSystem.isCaseSensitive() ? CASE_SENSITIVE : CASE_INSENSITIVE;
-            return new AtomicSnapshotHierarchyReference(DefaultSnapshotHierarchy.empty(caseSensitivity));
+            return new DefaultVfsRootReference(DefaultSnapshotHierarchy.empty(caseSensitivity), updateFunctionRunner);
         }
 
         VirtualFileSystem createVirtualFileSystem(
             FileHasher hasher,
-            AtomicSnapshotHierarchyReference root,
+            VfsRootReference rootReference,
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java:203]: Do we need to include `Reference` in the name?

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/FileWatcherUpdater.java [hunk 2]
```diff
@@ -17,13 +17,13 @@
...
 import org.gradle.internal.snapshot.CompleteFileSystemLocationSnapshot;
-import org.gradle.internal.snapshot.SnapshotHierarchy;
+import org.gradle.internal.vfs.SnapshotDiffListener;
 import org.gradle.internal.watch.WatchingNotSupportedException;
 
 import java.io.File;
 import java.util.Collection;
 
-public interface FileWatcherUpdater extends SnapshotHierarchy.SnapshotDiffListener {
+public interface FileWatcherUpdater extends SnapshotDiffListener {
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/FileWatcherUpdater.java:26]: I think this shouldn't implement the listener. We should instead create a separate listener that calls this thing. I think we are actually doing that already, so this here is just an oversight.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java [hunk 3]
```diff
@@ -103,12 +99,12 @@ private void updateWatchRegistry(Consumer<FileWatcherRegistry> updateFunction) {
     }
 
     private void updateWatchRegistry(Consumer<FileWatcherRegistry> updateFunction, Runnable noWatchRegistry) {
-        root.update(currentRoot -> {
+        rootReference.update(root -> {
             if (watchRegistry == null) {
                 noWatchRegistry.run();
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java:104]: Shall we name this something like `updateWihtoutWatcherRegistry`?

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java [hunk 4]
```diff
@@ -131,17 +127,16 @@ public void beforeBuildFinished(boolean watchingEnabled) {
...
-                newRoot = handleWatcherRegistryEvents(newRoot, "for current build");
+            rootReference.update(root -> {
+                removeSymbolicLinks(root);
+                handleWatcherRegistryEvents(root, "for current build");
                 if (watchRegistry != null) {
-                    newRoot = withWatcherChangeErrorHandling(newRoot, () -> watchRegistry.getFileWatcherUpdater().buildFinished());
+                    withWatcherChangeErrorHandling(root, () -> watchRegistry.getFileWatcherUpdater().buildFinished());
                 }
-                printStatistics(newRoot, "retains", "till next build");
-                return newRoot;
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java:141]: Why did you change this to be in-place modification instead of the immutable stuff we had? I think the original transform-like behavior was cleaner.

File: subprojects/snapshots/src/main/java/org/gradle/internal/vfs/VfsRoot.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package org.gradle.internal.vfs;
+
+import org.gradle.internal.snapshot.MetadataSnapshot;
+import org.gradle.internal.snapshot.ReadOnlyVfsRoot;
+
+import javax.annotation.CheckReturnValue;
+
+public interface VfsRoot extends ReadOnlyVfsRoot {
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/vfs/VfsRoot.java:24]: I think this should be `VfsRoot` and `MutableVfsRoot`. But I'm not convinced it's good to have a mutable version like this.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java (hunk 4 above)
Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java:141]: I think the mutable stuff is better since then we always get notifications about changes to the root reference. And the client actually wants to mutate the root.

**Comment by lptr** [comment]
Let's refactor this way:

image

Google doc: https://docs.google.com/drawings/d/1ywzPf745ILglz5ZQjkLGdi_rK3ywdosH1ixNwwXUH4w/edit

**Comment by wolfs** [comment]
Replaced by https://github.com/gradle/gradle/pull/13975.
